src/main/scala/Core/control.scala

```diff
@@ -322,23 +336,88 @@ class Control(implicit val conf: FlexpretConfiguration) extends Module
   //val exe_brjmp = exe_valid && (exe_reg_jump || (exe_reg_branch && io.exe_br_cond))
 
   // Keep track of delay_until instruction.
+  val exe_expire_du_wu = io.exe_expire_du || io.exe_expire_wu
+  val exe_expire_ie_ee = io.exe_expire_ie || io.exe_expire_ee
+  val exe_expire = exe_expire_du_wu || exe_expire_ie_ee
+  
+  // At the end of DU, another DU instruction will be fetched because DU is a
+  // branch to self. We need to make sure we do not execute this DU, otherwise we
+  // will end in a loop.
+  // We do this by capturing the thread id that fecthed a DU, and marking it
+  // invalid when it otherwise would be executed.
+  // 
+  // We do the same procedure for WU, but in this case it fixes issues related
+  // to running WU with a trigger < the current time.
+  val du_expired = RegInit(VecInit(Seq.fill(conf.threads) { false.B }))
+  val wu_expired = RegInit(VecInit(Seq.fill(conf.threads) { false.B }))
+  val kill_du = WireInit(false.B)
+  val kill_wu = WireInit(false.B)
+  
+  val exe_reg_du = RegNext(dec_du.asBool)
+  val exe_reg_wu = RegNext(dec_wu.asBool)
+  
+  // Any thread may have its DU triggered at any given time. In most cases, only
+  // a single thread will have its DU triggered, but it is not unthinkable that
+  // several threads will be set off to trigger simultaneously. When 'flex' is used,
+  // the ordering of threads is configurable at run-time, so we must handle all
+  // combinations.
```

Comment:
What do you mean by having its DU triggered?

Reply:
I think I'll rewrite these comments based on what I wrote in the other question addressing this :)